pyproject.toml

```diff
@@ -235,8 +236,10 @@ ignore = [
     "Q000",  # double quotes found but single quotes preferred
     "Q001",  # single quote docstring found but double quotes preferred
     # flake8-return
+    "RET503",  # Missing explicit `return` at the end of function able to return value
     "RET504",  # unnecessary variable assignment before `return` statement
     "RET505",  # unnecessary `else` after `return` statement
+    "RET506",  # Unnecessary `else` after `raise` statement
```

Comment:
should this PR not just focus on the introduction of pre-commit, rather than also changing the lints?

Reply:
My objective is to limit the impact to the chore code to the minimum so instead of modifying sphinx sources I decided to add extra ignores. I think they could then be dropped 1 by 1 in follow-up PR when someone feel motivated enough. Again for stability reasons.

Reply:
even worth when you know that some of them are not compatible (D212 - D213)

Reply:
Thoughts @AA-Turner ?

Reply:
Note that any changes to linter will break the the current PR and branches. My take on this one is to use several PRs:
1. set up the structure (pre-commit, pinned linter version) without changing the code
2. sanitize ruff rules (one by one to avoid breaking everything)
3. integrate mypy to tox ? 
4. see if we can apply other rules (black, prettier)

This PR is dedicated to 1.

Reply:
> Note that I don't know how they were merged in the first place as SIM114 was manually ignored on many lines.

I think that's because the ruff check only checks the diff and not old places. So if you add correct stuff, it won't show up other errors. To that end, we could add `args: ['--diff', '--format=github']` to the pre-commit config. If you do that, this would however mean that old codebased won't be ruffed but only new code will (and I think you won't need extra ignores).